File: backend/api/schemas/creator_media.py
"""
Creator Media Library Schemas

Pydantic schemas for the unified Creator Media Library.
Supports all user-uploaded assets that can be injected into generation prompts.
"""

from datetime import datetime
from typing import Optional, List, Dict, Any, Literal
from pydantic import BaseModel, Field


# ============================================================================
# Asset Type Definitions
# ============================================================================

MediaAssetType = Literal[
    "logo",           # Brand logos
    "face",           # User faces for thumbnails
    "character",      # Character/avatar representations
    "game_skin",      # Game character skins
    "object",         # Props, items to include
    "background",     # Custom backgrounds
    "reference",      # Style reference images
    "overlay",        # Stream overlays
    "emote",          # Channel emotes
    "badge",          # Subscriber badges
    "panel",          # Channel panels
    "alert",          # Alert images
    "facecam_frame",  # Facecam borders
    "stinger",        # Transition animations
]

MEDIA_ASSET_TYPES: List[MediaAssetType] = [
    "logo", "face", "character", "game_skin", "object", "background",
    "reference", "overlay", "emote", "badge", "panel", "alert",
    "facecam_frame", "stinger"
]


# ============================================================================
# Type-Specific Metadata Schemas
# ============================================================================

class LogoMetadata(BaseModel):
    """Metadata for logo assets."""
    logo_variant: Optional[Literal["primary", "secondary", "icon", "monochrome", "watermark"]] = None
    transparent: bool = True
    brand_kit_id: Optional[str] = None


class FaceMetadata(BaseModel):
    """Metadata for face assets."""
    expression: Optional[str] = None  # happy, surprised, serious, etc.
    angle: Optional[Literal["front", "side", "three_quarter"]] = "front"
    processed: bool = False  # Background removed
    processed_url: Optional[str] = None


class CharacterMetadata(BaseModel):
    """Metadata for character/avatar assets."""
    style: Optional[str] = None  # anime, realistic, cartoon, pixel
    outfit: Optional[str] = None
    pose: Optional[str] = None
    color_scheme: Optional[List[str]] = None


class GameSkinMetadata(BaseModel):
    """Metadata for game skin assets."""
    game: Optional[str] = None  # fortnite, valorant, league, etc.
    character_name: Optional[str] = None
    skin_name: Optional[str] = None
    rarity: Optional[str] = None  # common, rare, epic, legendary


class ObjectMetadata(BaseModel):
    """Metadata for object/prop assets."""
    category: Optional[str] = None  # prop, item, weapon, food, etc.
    transparent: bool = True


class BackgroundMetadata(BaseModel):
    """Metadata for background assets."""
    style: Optional[str] = None  # gradient, scene, abstract, gaming
    mood: Optional[str] = None  # energetic, calm, dark, vibrant
    colors: Optional[List[str]] = None


class ReferenceMetadata(BaseModel):
    """Metadata for reference images."""
    source: Optional[str] = None  # youtube, twitch, custom
    video_id: Optional[str] = None
    original_creator: Optional[str] = None
    notes: Optional[str] = None


# ============================================================================
# Request Schemas
# ============================================================================

class UploadMediaRequest(BaseModel):
    """Request to upload a new media asset."""
    asset_type: MediaAssetType = Field(
        ...,
        description="Type of media asset"
    )
    display_name: str = Field(
        ...,
        min_length=1,
        max_length=100,
        description="Display name for the asset"
    )
    description: Optional[str] = Field(
        default=None,
        max_length=500,
        description="Optional description"
    )
    image_base64: str = Field(
        ...,
        description="Base64-encoded image data"
    )
    tags: Optional[List[str]] = Field(
        default=None,
        max_length=20,
        description="Tags for organization (max 20)"
    )
    is_favorite: bool = Field(
        default=False,
        description="Mark as favorite"
    )
    set_as_primary: bool = Field(
        default=False,
        description="Set as primary asset of this type"
    )
    remove_background: Optional[bool] = Field(
        default=None,
        description="Remove background from image. Defaults based on asset type (True for face/logo/character/object/emote/badge, False for background/reference/panel)"
    )
    metadata: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Type-specific metadata"
    )

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "asset_type": "face",
                    "display_name": "My Streaming Face",
                    "description": "Happy expression for thumbnails",
                    "image_base64": "iVBORw0KGgo...",
                    "tags": ["happy", "front-facing"],
                    "is_favorite": True,
                    "set_as_primary": True,
                    "remove_background": True,
                    "metadata": {"expression": "happy", "angle": "front"}
                }
            ]
        }
    }


class UpdateMediaRequest(BaseModel):
    """Request to update a media asset."""
    display_name: Optional[str] = Field(
        default=None,
        min_length=1,
        max_length=100
    )
    description: Optional[str] = Field(
        default=None,
        max_length=500
    )
    tags: Optional[List[str]] = Field(
        default=None,
        max_length=20
    )
    is_favorite: Optional[bool] = None
    is_primary: Optional[bool] = None
    metadata: Optional[Dict[str, Any]] = None


class ListMediaRequest(BaseModel):
    """Query parameters for listing media assets."""
    asset_type: Optional[MediaAssetType] = None
    tags: Optional[List[str]] = None
    favorites_only: bool = False
    search: Optional[str] = None
    limit: int = Field(default=50, ge=1, le=100)
    offset: int = Field(default=0, ge=0)
    sort_by: Literal["created_at", "updated_at", "usage_count", "display_name"] = "created_at"
    sort_order: Literal["asc", "desc"] = "desc"


# ============================================================================
# Response Schemas
# ============================================================================

class MediaAsset(BaseModel):
    """A single media asset."""
    id: str
    user_id: str
    asset_type: MediaAssetType
    display_name: str
    description: Optional[str] = None
    url: str
    storage_path: str
    thumbnail_url: Optional[str] = None
    processed_url: Optional[str] = None  # Background-removed version
    processed_storage_path: Optional[str] = None
    file_size: Optional[int] = None
    mime_type: Optional[str] = None
    width: Optional[int] = None
    height: Optional[int] = None
    tags: List[str] = []
    is_favorite: bool = False
    is_primary: bool = False
    has_background_removed: bool = False
    metadata: Dict[str, Any] = {}
    usage_count: int = 0
    last_used_at: Optional[str] = None
    created_at: str
    updated_at: str

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "id": "550e8400-e29b-41d4-a716-446655440000",
                    "user_id": "user-123",
                    "asset_type": "face",
                    "display_name": "My Streaming Face",
                    "description": "Happy expression for thumbnails",
                    "url": "https://storage.example.com/faces/user-123/asset.png",
                    "storage_path": "creator-media/user-123/face/asset.png",
                    "thumbnail_url": "https://storage.example.com/faces/user-123/asset_thumb.png",
                    "processed_url": "https://storage.example.com/faces/user-123/asset_processed.png",
                    "processed_storage_path": "creator-media/user-123/face/asset_processed.png",
                    "file_size": 245000,
                    "mime_type": "image/png",
                    "width": 512,
                    "height": 512,
                    "tags": ["happy", "front-facing"],
                    "is_favorite": True,
                    "is_primary": True,
                    "has_background_removed": True,
                    "metadata": {"expression": "happy", "angle": "front"},
                    "usage_count": 15,
                    "last_used_at": "2025-01-01T12:00:00Z",
                    "created_at": "2024-12-01T10:00:00Z",
                    "updated_at": "2025-01-01T12:00:00Z"
                }
            ]
        }
    }


class UploadMediaResponse(BaseModel):
    """Response after uploading a media asset."""
    asset: MediaAsset
    message: str = "Asset uploaded successfully"


class ListMediaResponse(BaseModel):
    """Response for listing media assets."""
    assets: List[MediaAsset]
    total: int
    limit: int
    offset: int
    has_more: bool


class MediaSummary(BaseModel):
    """Summary of media assets by type."""
    asset_type: MediaAssetType
    total_count: int
    favorite_count: int
    latest_upload: Optional[str] = None


class MediaLibrarySummaryResponse(BaseModel):
    """Full summary of user's media library."""
    summaries: List[MediaSummary]
    total_assets: int
    storage_used_bytes: int


class DeleteMediaResponse(BaseModel):
    """Response after deleting a media asset."""
    success: bool
    message: str


class BulkDeleteMediaResponse(BaseModel):
    """Response after bulk deleting media assets."""
    deleted_count: int
    failed_ids: List[str] = []
    message: str


# ============================================================================
# Prompt Injection Schemas
# ============================================================================

class MediaForPrompt(BaseModel):
    """Media asset formatted for prompt injection."""
    id: str
    asset_type: MediaAssetType
    display_name: str
    url: str
    metadata: Dict[str, Any] = {}
    
    def to_prompt_context(self) -> str:
        """Generate prompt context string for this asset."""
        context_parts = [f"[{self.asset_type.upper()}] {self.display_name}"]
        
        if self.asset_type == "face":
            if self.metadata.get("expression"):
                context_parts.append(f"Expression: {self.metadata['expression']}")
            if self.metadata.get("angle"):
                context_parts.append(f"Angle: {self.metadata['angle']}")
        elif self.asset_type == "character":
            if self.metadata.get("style"):
                context_parts.append(f"Style: {self.metadata['style']}")
        elif self.asset_type == "game_skin":
            if self.metadata.get("game"):
                context_parts.append(f"Game: {self.metadata['game']}")
            if self.metadata.get("character_name"):
                context_parts.append(f"Character: {self.metadata['character_name']}")
        
        return " | ".join(context_parts)


class SelectedMediaForGeneration(BaseModel):
    """Media assets selected for a generation request."""
    face_id: Optional[str] = None
    logo_id: Optional[str] = None
    character_id: Optional[str] = None
    background_id: Optional[str] = None
    reference_ids: Optional[List[str]] = None
    object_ids: Optional[List[str]] = None
    game_skin_id: Optional[str] = None
    
    def has_any(self) -> bool:
        """Check if any media is selected."""
        return any([
            self.face_id, self.logo_id, self.character_id,
            self.background_id, self.reference_ids, self.object_ids,
            self.game_skin_id
        ])

File: tsx/packages/api-client/src/types/creatorMedia.ts
/**
 * Creator Media Library Types
 * 
 * TypeScript interfaces for the unified media library.
 * All user-uploaded assets that can be injected into generation prompts.
 */

// ============================================================================
// Asset Type Definitions
// ============================================================================

// ============================================================================
// Constants
// ============================================================================

export const TOTAL_ASSET_LIMIT = 25;
export const MAX_PROMPT_INJECTION_ASSETS = 2;
export const ALLOWED_TIERS = ['pro', 'studio'] as const;

// Asset types that should have background removed by default
export const BG_REMOVAL_DEFAULT_TYPES: MediaAssetType[] = [
  'face', 'logo', 'character', 'object', 'emote', 'badge', 'game_skin'
];

// Asset types that should NEVER have background removed
export const BG_REMOVAL_EXCLUDED_TYPES: MediaAssetType[] = [
  'background', 'reference', 'panel', 'overlay', 'alert', 'facecam_frame', 'stinger'
];

export function shouldRemoveBackgroundByDefault(assetType: MediaAssetType): boolean {
  return BG_REMOVAL_DEFAULT_TYPES.includes(assetType);
}

export function canRemoveBackground(assetType: MediaAssetType): boolean {
  return !BG_REMOVAL_EXCLUDED_TYPES.includes(assetType);
}

export type MediaAssetType =
  | 'logo'
  | 'face'
  | 'character'
  | 'game_skin'
  | 'object'
  | 'background'
  | 'reference'
  | 'overlay'
  | 'emote'
  | 'badge'
  | 'panel'
  | 'alert'
  | 'facecam_frame'
  | 'stinger';

export const MEDIA_ASSET_TYPES: MediaAssetType[] = [
  'logo', 'face', 'character', 'game_skin', 'object', 'background',
  'reference', 'overlay', 'emote', 'badge', 'panel', 'alert',
  'facecam_frame', 'stinger'
];

export const ASSET_TYPE_LABELS: Record<MediaAssetType, string> = {
  logo: 'Logos',
  face: 'Faces',
  character: 'Characters',
  game_skin: 'Game Skins',
  object: 'Objects',
  background: 'Backgrounds',
  reference: 'References',
  overlay: 'Overlays',
  emote: 'Emotes',
  badge: 'Badges',
  panel: 'Panels',
  alert: 'Alerts',
  facecam_frame: 'Facecam Frames',
  stinger: 'Stingers',
};

export const ASSET_TYPE_DESCRIPTIONS: Record<MediaAssetType, string> = {
  logo: 'Brand logos (primary, secondary, icon, etc.)',
  face: 'User faces for thumbnail recreation',
  character: 'Character/avatar representations',
  game_skin: 'Game character skins',
  object: 'Props and items to include in generations',
  background: 'Custom backgrounds',
  reference: 'Style reference images',
  overlay: 'Stream overlays',
  emote: 'Channel emotes',
  badge: 'Subscriber badges',
  panel: 'Channel panels',
  alert: 'Alert images',
  facecam_frame: 'Facecam borders',
  stinger: 'Transition animations',
};

// ============================================================================
// Core Types
// ============================================================================

export interface MediaAsset {
  id: string;
  userId: string;
  assetType: MediaAssetType;
  displayName: string;
  description?: string | null;
  url: string;
  storagePath: string;
  thumbnailUrl?: string | null;
  processedUrl?: string | null;  // Background-removed version
  processedStoragePath?: string | null;
  fileSize?: number | null;
  mimeType?: string | null;
  width?: number | null;
  height?: number | null;
  tags: string[];
  isFavorite: boolean;
  isPrimary: boolean;
  hasBackgroundRemoved: boolean;
  metadata: Record<string, any>;
  usageCount: number;
  lastUsedAt?: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface MediaSummary {
  assetType: MediaAssetType;
  totalCount: number;
  favoriteCount: number;
  latestUpload?: string | null;
}

export interface MediaForPrompt {
  id: string;
  assetType: MediaAssetType;
  displayName: string;
  url: string;
  metadata: Record<string, any>;
}

// ============================================================================
// Request Types
// ============================================================================

export interface UploadMediaRequest {
  assetType: MediaAssetType;
  displayName: string;
  description?: string;
  imageBase64: string;
  tags?: string[];
  isFavorite?: boolean;
  setAsPrimary?: boolean;
  removeBackground?: boolean;  // null = use default based on asset type
  metadata?: Record<string, any>;
}

export interface UpdateMediaRequest {
  displayName?: string;
  description?: string;
  tags?: string[];
  isFavorite?: boolean;
  isPrimary?: boolean;
  metadata?: Record<string, any>;
}

export interface ListMediaParams {
  assetType?: MediaAssetType;
  tags?: string[];
  favoritesOnly?: boolean;
  search?: string;
  limit?: number;
  offset?: number;
  sortBy?: 'created_at' | 'updated_at' | 'usage_count' | 'display_name';
  sortOrder?: 'asc' | 'desc';
}

// ============================================================================
// Response Types
// ============================================================================

export interface MediaAccessResponse {
  hasAccess: boolean;
  tier: string;
  totalLimit: number;
  maxPerPrompt: number;
  upgradeMessage: string | null;
}

export interface UploadMediaResponse {
  asset: MediaAsset;
  message: string;
}

export interface ListMediaResponse {
  assets: MediaAsset[];
  total: number;
  limit: number;
  offset: number;
  hasMore: boolean;
}

export interface MediaLibrarySummaryResponse {
  summaries: MediaSummary[];
  totalAssets: number;
  storageUsedBytes: number;
}

export interface DeleteMediaResponse {
  success: boolean;
  message: string;
}

export interface BulkDeleteMediaResponse {
  deletedCount: number;
  failedIds: string[];
  message: string;
}

export interface AssetTypesResponse {
  types: MediaAssetType[];
  descriptions: Record<MediaAssetType, string>;
}

// ============================================================================
// Prompt Injection Types
// ============================================================================

export interface SelectedMediaForGeneration {
  faceId?: string;
  logoId?: string;
  characterId?: string;
  backgroundId?: string;
  referenceIds?: string[];
  objectIds?: string[];
  gameSkinId?: string;
}

// ============================================================================
// Type-Specific Metadata
// ============================================================================

export interface LogoMetadata {
  logoVariant?: 'primary' | 'secondary' | 'icon' | 'monochrome' | 'watermark';
  transparent?: boolean;
  brandKitId?: string;
}

export interface FaceMetadata {
  expression?: string;
  angle?: 'front' | 'side' | 'three_quarter';
  processed?: boolean;
  processedUrl?: string;
}

export interface CharacterMetadata {
  style?: string;
  outfit?: string;
  pose?: string;
  colorScheme?: string[];
}

export interface GameSkinMetadata {
  game?: string;
  characterName?: string;
  skinName?: string;
  rarity?: string;
}

export interface ObjectMetadata {
  category?: string;
  transparent?: boolean;
}

export interface BackgroundMetadata {
  style?: string;
  mood?: string;
  colors?: string[];
}

export interface ReferenceMetadata {
  source?: string;
  videoId?: string;
  originalCreator?: string;
  notes?: string;
}


// ============================================================================
// Asset Placement Types
// ============================================================================

/**
 * Anchor point for positioning reference
 */
export type PositionAnchor = 'center' | 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right';

/**
 * Size unit type
 */
export type SizeUnit = 'percent' | 'px';

/**
 * Position data for an asset on the canvas
 */
export interface AssetPosition {
  /** X position as percentage (0-100) from left edge */
  x: number;
  /** Y position as percentage (0-100) from top edge */
  y: number;
  /** Reference point for the position */
  anchor: PositionAnchor;
}

/**
 * Size data for an asset on the canvas
 */
export interface AssetSize {
  /** Width value */
  width: number;
  /** Height value */
  height: number;
  /** Unit type - percentage of canvas or absolute pixels */
  unit: SizeUnit;
  /** Whether to maintain original aspect ratio when resizing */
  maintainAspectRatio: boolean;
}

/**
 * Complete placement data for a single asset
 */
export interface AssetPlacement {
  /** Reference to the media asset */
  assetId: string;
  /** The media asset data */
  asset: MediaAsset;
  /** Position on canvas */
  position: AssetPosition;
  /** Size on canvas */
  size: AssetSize;
  /** Layer order (higher = on top) */
  zIndex: number;
  /** Rotation in degrees (0-360) */
  rotation: number;
  /** Opacity (0-100) */
  opacity: number;
}

/**
 * Serialized placement data for API transmission
 */
export interface SerializedPlacement {
  assetId: string;
  displayName: string;
  assetType: MediaAssetType;
  url: string;
  x: number;
  y: number;
  width: number;
  height: number;
  sizeUnit: SizeUnit;
  zIndex: number;
  rotation: number;
  opacity: number;
}

/**
 * Serialize placements for API transmission
 */
export function serializePlacements(placements: AssetPlacement[]): SerializedPlacement[] {
  return placements.map(p => ({
    assetId: p.assetId,
    displayName: p.asset.displayName,
    assetType: p.asset.assetType,
    url: p.asset.processedUrl || p.asset.url,
    x: p.position.x,
    y: p.position.y,
    width: p.size.width,
    height: p.size.height,
    sizeUnit: p.size.unit,
    zIndex: p.zIndex,
    rotation: p.rotation,
    opacity: p.opacity,
  }));
}
